controllers/provider_controller.go

```diff
@@ -76,96 +79,73 @@ func (r *ProviderReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts P
 
 // +kubebuilder:rbac:groups=notification.toolkit.fluxcd.io,resources=providers,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=notification.toolkit.fluxcd.io,resources=providers/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
 
 func (r *ProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, retErr error) {
-	start := time.Now()
+	reconcileStart := time.Now()
 	log := ctrl.LoggerFrom(ctx)
 
-	provider := &v1beta1.Provider{}
-	if err := r.Get(ctx, req.NamespacedName, provider); err != nil {
+	obj := &apiv1.Provider{}
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	r.RecordSuspend(ctx, provider, provider.Spec.Suspend)
-	// return early if the object is suspended
-	if provider.Spec.Suspend {
-		log.Info("Reconciliation is suspended for this object")
-		return ctrl.Result{}, nil
-	}
-
-	patchHelper, err := patch.NewHelper(provider, r.Client)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+	// Initialize the runtime patcher with the current version of the object.
+	patcher := patch.NewSerialPatcher(obj, r.Client)
 
 	defer func() {
-		patchOpts := []patch.Option{
-			patch.WithOwnedConditions{
-				Conditions: []string{
-					meta.ReadyCondition,
-					meta.ReconcilingCondition,
-					meta.StalledCondition,
-				},
-			},
-		}
-
-		if retErr == nil && (result.IsZero() || !result.Requeue) {
-			conditions.Delete(provider, meta.ReconcilingCondition)
-
-			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
-
-			readyCondition := conditions.Get(provider, meta.ReadyCondition)
-			switch readyCondition.Status {
-			case metav1.ConditionFalse:
-				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
-				conditions.MarkStalled(provider, readyCondition.Reason, readyCondition.Message)
-			case metav1.ConditionTrue:
-				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
-				conditions.Delete(provider, meta.StalledCondition)
-			}
-		}
-
-		if err := patchHelper.Patch(ctx, provider, patchOpts...); err != nil {
-			retErr = kerrors.NewAggregate([]error{retErr, err})
+		// Record Prometheus metrics.
+		r.Metrics.RecordReadiness(ctx, obj)
+		r.Metrics.RecordDuration(ctx, obj, reconcileStart)
+		r.Metrics.RecordSuspend(ctx, obj, obj.Spec.Suspend)
+
+		// Issue warning event if the reconciliation failed.
+		if retErr != nil {
+			r.Event(obj, corev1.EventTypeWarning, meta.FailedReason, retErr.Error())
 		}
 
-		r.Metrics.RecordReadiness(ctx, provider)
-		r.Metrics.RecordDuration(ctx, provider, start)
-
+		// Patch finalizers, status and conditions.
+		retErr = r.patch(ctx, obj, patcher)
```

Comment:
Please correct me if I'm misunderstanding the implications of this.
I created https://go.dev/play/p/jTu7aBvKYaU to see if a defer would overwrite the named return value and looks like it does. Which means, an error would be replaced by nil if this deferred patch succeeds.

Reply:
That what mean all refactored controllers are no longer returning errors:
https://github.com/fluxcd/source-controller/blob/main/controllers/gitrepository_controller.go#L203
https://github.com/fluxcd/kustomize-controller/blob/main/controllers/kustomization_controller.go#L178

If this were true, then how come the `TestReceiverReconciler_Reconcile` passes?

Reply:
Reading the Reconcile tests, I think it passes because the error is returned with `Requeue: true` Result https://github.com/fluxcd/notification-controller/blob/6261b17928b3a294f26213938faab5bc889206bb/controllers/receiver_controller.go#L156 , which results in an immediate reconciliation, even if the error is overwritten as nil at the end.

In case of source controller, the SummarizeAndPatch process takes into consideration the Result and error, and returns the final result of reconciliation https://github.com/fluxcd/source-controller/blob/3fee9a57bf2ceaf7eacf7e6fa902be5abbaa6413/controllers/gitrepository_controller.go#L194 . This is why we always consider the error and the Result at the end of reconciliation for the final result of reconciliation.

In case of kustomize-controller, looks like this was introduced in the recent refactor.

Reply:
Ahh right, it works because of `Requeue: true`. Going to fix this in kustomize-controller too, great catch!